### PR TITLE
Allow implicit flattening for yaml inputs.

### PR DESCRIPTION
**Pull request by robertwb**
One can still explicitly use `Flatten` if desired.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @bzablocki

**Review by bzablocki — COMMENTED** [review]
Thanks @robertwb. 
I have one other, general question, which probably was discussed multiple times already. Why aren't type hints used in the beam's codebase? Is there a case against it or is it just a legacy thing?

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 1]
```diff
@@ -250,29 +260,35 @@ def expand_transform(spec, scope):
...
-    inputs = inputs_dict
+  ptransform = scope.create_ptransform(spec)
+  transform_label = scope.unique_name(spec, ptransform)
+
+  if spec['type'] == 'Flatten':
+    # Avoid flattening before the flatten, just to make a nicer graph.
+    inputs = tuple(
+        scope.get_pcollection(input) for key,
+        value in spec['input'].items()
+        for input in ([value] if isinstance(value, str) else value))
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/yaml/yaml_transform.py]: Ah, yeah. Yapf has trouble with the commas in list comprehensions. Fixed by adding parentheses.

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 2]
```diff
@@ -123,6 +123,16 @@ def compute_all(self):
...
 
+  def get_input(self, inputs, transform_label, key):
+    if isinstance(inputs, str):
+      return self.get_pcollection(inputs)
+    else:
+      return tuple(
+          self.get_pcollection(x) for x in
+          inputs) | f'{transform_label}-FlattenInputs[{key}]' >> beam.Flatten(
+              pipeline=self.root if isinstance(self.root, beam.Pipeline
+                                               ) else self.root.pipeline)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/yaml/yaml_transform.py]: That pipeline statement is certainly unwieldy. Refactored a bit, hopefully it's clearer now.

**Comment by robertwb** [comment]
Regarding type hints, this is mostly because the codebase predates PEP 484 (and Python 3). We have added type hints in some places, and do run mypy. It would probably be a good project to add type hints here (and it's probably self-contained enough that it wouldn't be a herculean project of fixing the rest of the SDK to do so).

**Review by bzablocki — APPROVED** [review]
Awesome, thanks!

**Pull request merged**